_posts/testing-wai.md

```diff
@@ -0,0 +1,222 @@
+
+In my previous [article](https://wasmer.io/posts/WAI-is-the-answer), I explained how to port a package from Rust for the JavaScript ecosystem. These published packages need testing in their own ecosystems for the package verification and identification of any possible bugs.
+
+## Testing in WAI
+
+Testing for WAI has been made easier using an automated testing framework `[wasmer-pack-testing](https://crates.io/crates/wasmer-pack-testing)`. This crate automatically discovers the tests for available language configurations for WAI and runs the tests correspondingly.
+
+It uses `jest` for JavaScript/Typescript and `pytest` for Python.
+
+This testing framework allows for both **logical checks** using unit tests and **code generation** using snapshot testing.
+
+## Tutorial Outline
+
+- Introduction
+- Environment Setup
+- Bindings generation
+- Testing for JavaScript
+
+### Introduction
+
+In this tutorial, we will test the recently published `rustfft` package. This package demonstrates the portability and logical computation power of WebAssembly that can be harnessed using WAI.
+
+<aside style="background: #fef08a; border-radius: 0.4rem; padding: 4px; font-weight: 500;">
+<span style="background: white; padding: 4px; border-radius: 9999px; margin: 0 2px 0 0">💡</span> RustFFT is a high-performance FFT library written in pure Rust.
+</aside>
+
+### Environment Setup
+
+The default testing using `cargo test` only runs the tests that are written using the `test` macro in the library files. For running integration tests we can create a new file in a separate tests directory.
+
+1. Create an integration tests file
+
+   ```bash
+   $ mkdir tests && touch tests/rustfft-integration-tests.rs
+   ```
+
+2. Let’s update our `cargo.toml` to tell it that we are providing a custom test harness.
+
+   ```toml
+   # Cargo.toml
+   [dev-dependencies]
+   anyhow = "1.0.68"
+   tracing-subscriber = { version = "0.3.16", features = ["env-filter"] }
+   wasmer-pack-testing = "0.7.0"
+
+   [[test]]
+   name = "rustfft-integration-tests"
+   harness = false
+   ```
+
+3. Finally, set up the integration tests to generate bindings and test for the appropriate target.
+
+   ```rust
+   // tests/sgp4-integration-tests.rs
+   use anyhow::Error;
+   use tracing_subscriber::EnvFilter;
+
+   fn main() -> Result<(), Error> {
+       tracing_subscriber::fmt()
+           .with_env_filter(EnvFilter::from_default_env())
+           .init();
+       wasmer_pack_testing::autodiscover(env!("CARGO_MANIFEST_DIR"))?;
+       Ok(())
+   }
+   ```
+
+### Bindings Generation
+
+For our testing suite to discover which targets to generate bindings for let’s create a blank JavaScript test file in our crate directory (the one with `Cargo.toml`) that `autodiscover()` can use.
+
+Now let’s create a blank `rustfft.test.js` in our main directory.
```

Comment:
By main directory here, do you mean the create directory from the other line?

Reply:
yeah